text_encoder.go

```diff
@@ -0,0 +1,230 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"sync"
+	"time"
+)
+
+var textPool = sync.Pool{New: func() interface{} {
+	return &textEncoder{
+		bytes: make([]byte, 0, _initialBufSize),
+	}
+}}
+
+type textEncoder struct {
+	bytes       []byte
+	timeFmt     string
+	firstNested bool
+}
+
+// NewTextEncoder creates a line-oriented text encoder whose output is optimized
+// for human, rather than machine, consumption. By default, the encoder uses
+// RFC3339-formatted timestamps.
+func NewTextEncoder(options ...TextOption) Encoder {
+	enc := textPool.Get().(*textEncoder)
+	enc.truncate()
+	enc.timeFmt = time.RFC3339
+	for _, opt := range options {
+		opt.apply(enc)
+	}
+	return enc
+}
+
+func (enc *textEncoder) Free() {
+	textPool.Put(enc)
+}
+
+func (enc *textEncoder) AddString(key, val string) {
+	enc.addKey(key)
+	enc.bytes = append(enc.bytes, val...)
+}
+
+func (enc *textEncoder) AddBool(key string, val bool) {
+	enc.addKey(key)
+	if val {
+		enc.bytes = append(enc.bytes, "true"...)
+		return
+	}
+	enc.bytes = append(enc.bytes, "false"...)
+}
+
+func (enc *textEncoder) AddInt(key string, val int) {
+	enc.AddInt64(key, int64(val))
+}
+
+func (enc *textEncoder) AddInt64(key string, val int64) {
+	enc.addKey(key)
+	enc.bytes = strconv.AppendInt(enc.bytes, val, 10)
+}
+
+func (enc *textEncoder) AddUint(key string, val uint) {
+	enc.AddUint64(key, uint64(val))
+}
+
+func (enc *textEncoder) AddUint64(key string, val uint64) {
+	enc.addKey(key)
+	enc.bytes = strconv.AppendUint(enc.bytes, val, 10)
+}
+
+func (enc *textEncoder) AddFloat64(key string, val float64) {
+	enc.addKey(key)
+	switch {
+	case math.IsNaN(val):
+		enc.bytes = append(enc.bytes, "NaN"...)
+	case math.IsInf(val, 1):
+		enc.bytes = append(enc.bytes, "+Inf"...)
+	case math.IsInf(val, -1):
+		enc.bytes = append(enc.bytes, "-Inf"...)
+	default:
+		enc.bytes = strconv.AppendFloat(enc.bytes, val, 'f', -1, 64)
+	}
+}
+
+func (enc *textEncoder) AddMarshaler(key string, obj LogMarshaler) error {
+	enc.addKey(key)
+	enc.firstNested = true
+	enc.bytes = append(enc.bytes, '{')
+	err := obj.MarshalLog(enc)
+	enc.bytes = append(enc.bytes, '}')
```

Comment:
We should reset `firstNested` to `false` here. I think a failing test case for this would:
 - add an object where the `MarhsalLog` adds no fields
 - then add a normal field (like string)

So we'd want something like `emptyObj={} something=val` but we might get `emptyObj={}something=val`